ci/ci/templates/index.html

```diff
@@ -118,6 +119,16 @@ <h3>PRs</h3>
           <td align="left">
             {{ pr.author }}
           </td>
+          <td align="left">
+            {% if 'failure' == pr.build_state %}
+            <form action="/invalidate_and_retest" method="post">
+              <input type="hidden" name="branch_id" value="{{ wb.branch }}"/>
+              <input type="hidden" name="pr_number" value="{{ pr.number }}"/>
+              <input type="hidden" name="_csrf" value="{{ token }}"/>
```

Comment:
You need to add the token to the context in index.html. This is the example from batch:

```
@prom_async_time(REQUEST_TIME_GET_BATCHES_UI)
@routes.get('/batches', name='batches')
@web_authenticated_users_only
async def ui_batches(request, userdata):
    params = request.query
    user = userdata['username']
    batches = await _get_batches_list(params, user)
    token = new_csrf_token()
    context = {'batch_list': batches, 'token': token}

    response = aiohttp_jinja2.render_template('batches.html',
                                              request,
                                              context)
    response.set_cookie('_csrf', token, secure=True, httponly=True)
    return response
```

Reply:
I used the authorize_source_sha endpoint for inspiration. Is that wrong too?